docs/specs/src/hwp_4_formal.md

```diff
@@ -0,0 +1,534 @@
+
+**Fundamental Write Constraint:** The DHT can never be "written" to directly. All data enters the DHT **only** by having been committed to an agent's source chain and then being **transformed** from validated local chain state into the elements (DHT operations) required for GDHT representation and lookup.
+
+**Structure of GDHT data:** The DHT is a content-addressable space where each piece of content is found at the address which is the hash of its content. In addition, any address in the DHT can have metadata attached to it. This metadata is not part of the content being hashed.
+
+*Note about hashing: Holochain uses 256-bit Blake2b hashes with the exception of one entry type, AgentPubKey, which is a 256-bit Ed25519 public key and uses `equal()` as its hash function. In other words, **the content of the AgentPubKey is identical to its hash**. This preserves content-addressability but also enables agent keys to function as self-proving identifiers for network transport and cryptographic functions like signing or encryption.*
+
+**DHT Addresses:** Both Actions and Entries from source chains can be retrieved from the DHT by either the ActionHash or EntryHash. The DHT `get()` function call returns a Record, a tuple containing the most relevant action/entry pair. Structurally, Actions "contain" their referenced entries so that pairing is obvious when a Record is retrieved by ActionHash. However, Actions are also attached as metadata at an EntryHash, and there could be many Actions which have created the same Entry content. A `get()` function called by EntryHash returns the oldest undeleted Action in the pair, while a `get_details()` function call on an EntryHash returns all of the Actions.
+
+**Agent Addresses & Agent Activity:** Technically an AgentPubKey functions as both a content address (which is never really used because performing a `get()` on the key just returns the key you used to get it) and a network address to send communications to that agents. But in addition to the content of the key stored on the DHT is metadata about that agent's chain activity. In other words, a `get_agent_activity()` request retrieves metadata about their chain records and chain status.
+
+Formally, the entire GDHT is represented as a set of 'basis hashes' $b_{c_x}$, or addresses where both content $c$ and metadata $m$ may be stored:
+
+$$
+\begin{aligned}
+GDHT &= \{d_1, \dots, d_n \} \\
+d_{b_{c_x}} &= \{c_x, M \} \\
+b_{c_x} &= hash(c_x) \\
+C &= E \bigsqcup H \\
+E &= \{e_1, \dots, e_n\} \\
+H &= \{h_1, \dots, h_n\} \\
+M &= \{m_1, \dots, m_n \} \\
+m_x &= \text{metadata}
+\end{aligned}
+$$
+
+There may be arbitrary types of metadata. Some important ones are:
+
+$$
+\begin{aligned}
+\forall e \: M_{context} &= \{{h_1}_e, \dots, {h_n}_e \} \\
+\exists c : M_{link} &= \{ link_1, \dots, link_n \} \\
+link &= \{ type, tag, b_{c_T} \} \\
+\end{aligned}
+$$
```

Comment:
The notation here is a bit loose, and I think it would be clearer to either tighten it up, or do away with it and just explain this part in English.

Problems I see with the notation:
- a new subscript is added to $h$ without explaining what it means. It could be good to formalize the [surjecive](https://en.wikipedia.org/wiki/Surjective_function) (many-to-one) relationship between H and E in the definitions above, making it clear that every H is the H of some E and introducing the notation for that there.
- the $\exists c$ should be attached to $link$, not $M_link$, since the $b_c_T$ of each link is different.
- although it should probably actually say $\exists T$ instead, since the notation above says $c_x$ is the content at index $x$, i.e. $x$ is the changing part, not $c$